{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "1.first read the sequence file and convert into dictionary- species with sequence\n",
    "Should I use class (?)\n",
    "2.To generate the plot: \n",
    "2.1 Species' name(not for id) in plot left side\n",
    "2.2 Sequence on the right side\n",
    "2.3 Compare sequence (column) for the most common same one use one color\n",
    "2.4 For other not the same amino acid with the most common one use other colors, each color dif\n",
    "2.5 Go through each column and generate the plot\n",
    "2.6 Some settings for plot\n",
    "3. Under the plot,\n",
    "3.1 Generate the Web Logo \n",
    "4. Split the protein sequence \n",
    "4.1 Compare Again and generate the table kind of result to show the similarity (local alignment)\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1.first read the sequence file and convert into dictionary- species with sequence\n",
    "my_dict = {}\n",
    "f = open(\"fzd_whole.fasta\",\"r\")\n",
    "def read_file_into_dic (filename):\n",
    "    \n",
    "    lines = f.readlines()\n",
    "    for line in lines:\n",
    "        if line.startswith(\">\"):\n",
    "            spec_name_only = line.split('[')[-1].split(']')[0].strip()\n",
    "            sequences = ''\n",
    "        else:\n",
    "            sequences += line.strip()\n",
    "            my_dict[spec_name_only] = sequences\n",
    "    return my_dict\n",
    "read_file_into_dic(f)\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "extracted_sequences = {species: sequence[28:150] for species, sequence in my_dict.items()}\n",
    "extracted_sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "values = list(extracted_sequences.values())\n",
    "print(values)\n",
    "type(values)\n",
    "# value is a list, list has two strings rn\n",
    "\n",
    "    # seq1=values[0]\n",
    "    # seq1 is whole string\n",
    "    # seq2=values[1]\n",
    "    # seq2 is whole string\n",
    "    \n",
    "amino_acid_list = [list(seq) for seq in values]\n",
    "print(amino_acid_list[0][2])\n",
    "\n",
    "\n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "\n",
    "aa_to_color = {\n",
    "    'A': '#FF0000',  \n",
    "    'C': '#00FF00',  \n",
    "    'D': '#0000FF',  \n",
    "    'E': '#FFFF00',  \n",
    "    'F': '#FF00FF',  \n",
    "    'G': '#00FFFF',  \n",
    "    'H': '#800000',  \n",
    "    'I': '#808000',  \n",
    "    'K': '#008000',  \n",
    "    'L': '#800080',  \n",
    "    'M': '#008080',  \n",
    "    'N': '#000080',  \n",
    "    'P': '#FFA500',  \n",
    "    'Q': '#A52A2A', \n",
    "    'R': '#808080',  \n",
    "    'S': '#C0C0C0',  \n",
    "    'T': '#FFFFFF',  \n",
    "    'V': '#000000', \n",
    "    'W': '#FF6347',  \n",
    "    'Y': '#4682B4',  \n",
    "    '-': '#D3D3D3',  \n",
    "}\n",
    "\n",
    "\n",
    "aa_matrix = np.array(amino_acid_list)\n",
    "color_matrix = np.array([[aa_to_color[aa] for aa in seq] for seq in amino_acid_list])\n",
    "\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(len(aminoacid_list[0]) * 0.7, len(amino_acid_list) * 0.7))\n",
    "\n",
    "\n",
    "for i in range(len(aminoacid_list)):\n",
    "    for j in range(len(aminoacid_list[0])):\n",
    "        ax.add_patch(plt.Rectangle((j, i), 1, 1, color=color_matrix[i][j]))\n",
    "        ax.text(j + 0.5, i + 0.5, aa_matrix[i, j], ha='center', va='center', fontsize=6)\n",
    "\n",
    "\n",
    "ax.set_yticks(np.arange(len(aminoacid_list)) + 0.5)\n",
    "ax.set_yticklabels([f'Seq {i+1}' for i in range(len(aminoacid_list))], fontsize=10)\n",
    "\n",
    "\n",
    "ax.set_xticks(np.arange(len(aminoacid_list[0])) + 0.5)\n",
    "ax.set_xticklabels(np.arange(1, len(aminoacid_list[0]) + 1), fontsize=6)\n",
    "\n",
    "\n",
    "ax.set_xlim(0, len(aminoacid_list[0]))\n",
    "ax.set_ylim(0, len(aminoacid_list))\n",
    "\n",
    "\n",
    "ax.set_xlabel('Position')\n",
    "ax.set_ylabel('Sequence')\n",
    "plt.title('MSA Visualization')\n",
    "\n",
    "\n",
    "ax.tick_params(left=False, bottom=False, labelleft=True, labelbottom=True)\n",
    "ax.grid(False)\n",
    "\n",
    "\n",
    "plt.tight_layout()\n",
    "\n",
    "\n",
    "plt.show()\n"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
